Mathlib/Order/SuccPred/Inverse.lean

```diff
@@ -0,0 +1,344 @@
+connection with `succ`.
+
+We also dually define `succ'` for a `PredOrder`.
+-/
+
+namespace Order
+
+open Set OrderDual
+
+variable {α : Type*}
+
+/-! ### Predecessor from successor -/
+
+
+open Classical in
+/-- A predecessor-like operation defined in a succesor order.
+
+If `a` is the succesor of some non-maximal `b`, `pred' a` returns an arbitrary such `b`. Otherwise,
+`pred' a = a`. -/
+noncomputable def pred' [Preorder α] [SuccOrder α] (a : α) : α :=
+  if h : ∃ b, ¬ IsMax b ∧ succ b = a then Classical.choose h else a
+
+section Preorder
+
+variable [Preorder α] [SuccOrder α] {a b : α}
+
+theorem pred'_ne_self_iff' : pred' a ≠ a ↔ ∃ b, ¬ IsMax b ∧ succ b = a := by
+  constructor <;>
+  intro H
+  · by_contra ha
+    rw [pred', dif_neg ha] at H
+    exact H.irrefl
+  · rw [pred', dif_pos H]
+    obtain ⟨hb, hb'⟩ := Classical.choose_spec H
+    conv_rhs => rw [← hb']
+    exact (lt_succ_of_not_isMax hb).ne
+
+theorem pred'_eq_self_iff' : pred' a = a ↔ ∀ b, succ b = a → IsMax b := by
+  simp_rw [← not_ne_iff, pred'_ne_self_iff', not_exists, not_and', not_not]
+
+/-- See `pred'_eq_self_iff'` for the most general version. -/
+theorem pred'_eq_self (ha : a ∉ range succ) : pred' a = a :=
+  pred'_eq_self_iff'.2 fun b h => (ha ⟨b, h⟩).elim
+
+theorem not_isMax_pred'_succ (ha : ¬ IsMax a) : ¬ IsMax (pred' (succ a)) := by
+  have H : ∃ b, ¬ IsMax b ∧ succ b = succ a := ⟨_, ha, rfl⟩
+  rw [pred', dif_pos H]
+  exact (Classical.choose_spec H).1
+
+@[simp]
+theorem succ_pred'_succ_of_not_isMax (ha : ¬ IsMax a) : succ (pred' (succ a)) = succ a := by
+  have H : ∃ b, ¬ IsMax b ∧ succ b = succ a := ⟨_, ha, rfl⟩
+  rw [pred', dif_pos H]
+  exact (Classical.choose_spec H).2
+
+theorem pred'_le_self (a : α) : pred' a ≤ a := by
+  by_cases H : ∃ b, ¬ IsMax b ∧ succ b = a
+  · obtain ⟨b, hb, rfl⟩ := H
+    apply (le_succ _).trans
+    rw [succ_pred'_succ_of_not_isMax hb]
+  · rw [pred', dif_neg H]
+
+theorem succ_pred'_eq_self_iff_of_not_isMax (ha : ¬ IsMax a) :
+    succ (pred' a) = a ↔ a ∈ range succ := by
+  refine ⟨fun h ↦ ⟨_, h⟩, ?_⟩
+  rintro ⟨b, rfl⟩
+  rw [succ_pred'_succ_of_not_isMax]
+  obtain ⟨c, hc⟩ := not_isMax_iff.1 ha
+  exact ((le_succ b).trans_lt hc).not_isMax
+
+theorem pred'_wcovBy (a : α) : pred' a ⩿ a := by
+  by_cases H : ∃ b, ¬ IsMax b ∧ succ b = a
+  · obtain ⟨b, hb, rfl⟩ := H
+    conv_rhs => rw [← succ_pred'_succ_of_not_isMax hb]
+    exact wcovBy_succ _
+  · rw [pred', dif_neg H]
+
+section NoMaxOrder
+
+variable [NoMaxOrder α]
+
+theorem pred'_ne_self_iff : pred' a ≠ a ↔ a ∈ range succ := by
+  simp [pred'_ne_self_iff']
+
+theorem pred'_eq_self_iff : pred' a = a ↔ a ∉ range succ := by
+  rw [← not_iff_not, not_not_mem, ← pred'_ne_self_iff]
+
+theorem succ_pred'_succ : succ (pred' (succ a)) = succ a :=
+  succ_pred'_succ_of_not_isMax (not_isMax a)
+
+theorem succ_pred'_eq_self_iff : succ (pred' a) = a ↔ a ∈ range succ :=
+  succ_pred'_eq_self_iff_of_not_isMax (not_isMax a)
+
```

Comment:
One direction of `pred'_ne_self_iff`, `pred'_eq_self_iff` and `succ_pred'_eq_self_iff` is true without the `NoMaxOrder` assumption. For `pred'_eq_self_iff` that's `pred'_eq_self`, but could you add those lemmas for the other two?